packages/entity/src/EntityMutator.ts

```diff
@@ -78,7 +78,35 @@ abstract class BaseMutator<
     >,
     protected readonly databaseAdapter: EntityDatabaseAdapter<TFields>,
     protected readonly metricsAdapter: IEntityMetricsAdapter
-  ) {}
+  ) {
+    this.mutationTriggers = [companionProvider.globalMutationTriggers, mutationTriggers].reduce<
```

Comment:
Instead of doing the merge here (which runs every time we create a mutator), I think we can do the merge earlier in the `EntityCompanion` constructor, which runs just once per entity type.

```diff
       this.privacyPolicy,
       entityCompanionDefinition.mutationValidators ?? [],
-      entityCompanionDefinition.mutationTriggers ?? {},
+      mergeEntityMutationTriggerConfigurations([
+        entityCompanionDefinition.mutationTriggers ?? {},
+        entityCompanionProvider.globalMutationTriggers ?? {},
+      ]),
       this.entityLoaderFactory,
       tableDataCoordinator.databaseAdapter,
```

Then, for the merging, I've found it to be a bit easier to read to explicitly list all the properties (which I agree is verbose, but it's hard to express types of keys of interfaces well in typescript. in this repo, having strict types is critical since it's primarily how we do breaking changes):

```typescript
function nonNullish<TValue>(value: TValue | null | undefined): value is NonNullable<TValue> {
  return value !== null && value !== undefined;
}

function mergeEntityMutationTriggerConfigurations<
  TFields extends object,
  TID extends NonNullable<TFields[TSelectedFields]>,
  TViewerContext extends ViewerContext,
  TEntity extends ReadonlyEntity<TFields, TID, TViewerContext, TSelectedFields>,
  TSelectedFields extends keyof TFields,
>(
  mutationTriggerConfigurations: EntityMutationTriggerConfiguration<
    TFields,
    TID,
    TViewerContext,
    TEntity,
    TSelectedFields
  >[],
): EntityMutationTriggerConfiguration<TFields, TID, TViewerContext, TEntity, TSelectedFields> {
  return {
    beforeCreate: mutationTriggerConfigurations.flatMap((c) => c.beforeCreate).filter(nonNullish),
    afterCreate: mutationTriggerConfigurations.flatMap((c) => c.afterCreate).filter(nonNullish),
    beforeUpdate: mutationTriggerConfigurations.flatMap((c) => c.beforeUpdate).filter(nonNullish),
    afterUpdate: mutationTriggerConfigurations.flatMap((c) => c.afterUpdate).filter(nonNullish),
    beforeDelete: mutationTriggerConfigurations.flatMap((c) => c.beforeDelete).filter(nonNullish),
    afterDelete: mutationTriggerConfigurations.flatMap((c) => c.afterDelete).filter(nonNullish),
    beforeAll: mutationTriggerConfigurations.flatMap((c) => c.beforeAll).filter(nonNullish),
    afterAll: mutationTriggerConfigurations.flatMap((c) => c.afterAll).filter(nonNullish),
    afterCommit: mutationTriggerConfigurations.flatMap((c) => c.afterCommit).filter(nonNullish),
  };
}
```

Reply:
Agreed!